Kernel/Arch/aarch64/RPi/InterruptController.cpp

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2022, Timon Kruiper <timonkruiper@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Format.h>
+#include <Kernel/Arch/aarch64/RPi/InterruptController.h>
+#include <Kernel/Arch/aarch64/RPi/MMIO.h>
+#include <Kernel/Interrupts/GenericInterruptHandler.h>
+
+namespace Kernel::RPi {
+
+// "7.5 Interrupts Registers"
+struct InterruptControllerRegisters {
+    u32 IRQ_basic_pending;
+    u32 IRQ_pending_1;
+    u32 IRQ_pending_2;
+    u32 FIQ_control;
+
+    u32 enable_IRQs_1;
+    u32 enable_IRQs_2;
+    u32 enable_basic_IRQs;
+
+    u32 disable_IRQs_1;
+    u32 disable_IRQs_2;
+    u32 disable_basic_IRQs;
+};
+
+InterruptController::InterruptController()
+    : m_registers(MMIO::the().peripheral<InterruptControllerRegisters>(0xB200))
+{
+}
+
+void InterruptController::enable(GenericInterruptHandler const& handler)
+{
+    u8 interrupt_number = handler.interrupt_number();
+    VERIFY(interrupt_number < 64);
+
+    if (interrupt_number < 32)
+        m_registers->enable_IRQs_1 = m_registers->enable_IRQs_1 | (1 << interrupt_number);
+    else
+        m_registers->enable_IRQs_2 = m_registers->enable_IRQs_2 | (1 << (interrupt_number - 32));
+}
+
+void InterruptController::disable(GenericInterruptHandler const& handler)
+{
+    u8 interrupt_number = handler.interrupt_number();
+    VERIFY(interrupt_number < 64);
+
+    if (interrupt_number < 32)
+        m_registers->disable_IRQs_1 = m_registers->disable_IRQs_1 | (1 << interrupt_number);
+    else
+        m_registers->disable_IRQs_2 = m_registers->disable_IRQs_2 | (1 << (interrupt_number - 32));
+}
+
+void InterruptController::eoi(GenericInterruptHandler const&) const
+{
+    // NOTE: The interrupt should be cleared by the device itself.
```

Comment:
cleared by the device? or by the device driver?

Reply:
I clarified the comment:
```
    // NOTE: The interrupt controller cannot clear the interrupt, since it is basically just a big multiplexer.
    //       The interrupt should be cleared by the corresponding device driver, such as a timer or uart.
```